### ISSUE TITLE
Inference?

### ISSUE THREAD

**Issue by leannmlindsey** [kuleshov-group/caduceus#49]
Yair,

Have you released any information about how to use the fine tuned models for inference?
L

**Comment by yair-schiff** [comment]
Hi @leannmlindsey, can you please clarify the question? I am not sure what you mean by how to use the models for inference. I guess it would depend on the task. There is code in this repo already for loading a pre-trained backbone and either fine-tuning a prediction head on top of embeddings or dumping the embeddings and training a separate model on top of those frozen embeddings.

**Comment by leannmlindsey** [comment]
I have several models that I have pre-trained using your pretraining instructions and then fine tuned using the [genomic benchmark code](https://github.com/kuleshov-group/caduceus/blob/main/slurm_scripts/run_genomics_benchmark.sh), but replacing the datasets with my own datasets.  

Unfortunately, I have been unable to find a way to properly load the checkpoint and configuration from the saved issues when trying to load trained Caduceus models for inference. Here is a list of the issues that I have encountered: 

To be clear, I am trying to load a model from the checkpoint/last.ckpt and model_config.json files that are saved in the output directory for a fine-tuned model using the given code from [genomic benchmark code](https://github.com/kuleshov-group/caduceus/blob/main/slurm_scripts/run_genomics_benchmark.sh).

1. Architecture Mismatch:
   - The checkpoint contains weights for ~64 layers
   - The model name suggests a 4-layer architecture ("4L" in caduceus_ps_char_4k_d256_4L)
   - This fundamental mismatch prevents proper weight loading

2. Configuration File Ambiguity:
   - The repository contains both config.json and model_config.json without documentation of their purposes
   - Neither appears to match the architecture of the saved checkpoint

3. State Dictionary Key Mismatches:
   - Checkpoint keys include: "model.caduceus.backbone.layers.X.mixer.submodule.mamba_fwd"
   - Model expects: "caduceus.backbone.layers.X.mixer.mamba_fwd"
   - These naming inconsistencies prevent automatic weight loading

4. Complex Infrastructure Requirements:
   - Need to handle dataset configuration even when just loading model for inference
   - Required understanding of PyTorch Lightning setup
   - No clear documentation for inference-only usage
   - Dependencies on training infrastructure for simple inference tasks

Feature Request:
For better usability, please consider implementing a simple interface for loading trained models:

```python
from caduceus import load_pretrained_model
model = load_pretrained_model(checkpoint_path)
model.eval()
predictions = model(input_sequence)

In the meantime, I will try your method of dumping the embeddings and training a separate model on top of the frozen embeddings.

**Comment by JiaLonghao1997** [comment]
> Hi @leannmlindsey, can you please clarify the question? I am not sure what you mean by how to use the models for inference. I guess it would depend on the task. There is code in this repo already for loading a pre-trained backbone and either fine-tuning a prediction head on top of embeddings or dumping the embeddings and training a separate model on top of those frozen embeddings.

After pre-training using the following code, `outputs/2024-12-23/02-38-12-125655/checkpoints/last.ckpt` was generated.  My question is how to use the generated model weights `checkpoints/last.ckpt` to perform model inference and obtain the embeddings of the new sequence.

```python
python -m train \
  experiment=hg38/hg38 \
  callbacks.model_checkpoint_every_n_steps.every_n_train_steps=500 \
  dataset.max_length=1024 \
  dataset.batch_size=1024 \
  dataset.mlm=true \
  dataset.mlm_probability=0.15 \
  dataset.rc_aug=false \
  model=caduceus \
  model.config.d_model=128 \
  model.config.n_layer=4 \
  model.config.bidirectional=true \
  model.config.bidirectional_strategy=add \
  model.config.bidirectional_weight_tie=true \
  model.config.rcps=true \
  optimizer.lr="8e-3" \
  train.global_batch_size=1024 \
  trainer.max_steps=10000 \
  +trainer.val_check_interval=10000 \
  wandb=null
```

**Comment by leannmlindsey** [comment]
@JiaLonghao1997  - I was able to eventually figure out the correct way to load the fine tuned model and do inference. I can send you my code in a day or two (after the holidays). The main issue was that I had to load from the fine tuned checkpoint configuration, not the model configuration (as the model configuration is set up for just the backbone). Once I figured that out, then inference worked.

**Comment by JiaLonghao1997** [comment]
> @JiaLonghao1997 - I was able to eventually figure out the correct way to load the fine tuned model and do inference. I can send you my code in a day or two (after the holidays). The main issue was that I had to load from the fine tuned checkpoint configuration, not the model configuration (as the model configuration is set up for just the backbone). Once I figured that out, then inference worked.

Thank you for your reply and Merry Christmas! I would really appreciate it if you could share your code.

**Comment by leannmlindsey** [comment]
@JiaLonghao1997 You can find my code for batch inference [here](https://github.com/leannmlindsey/caduceus). There is a slurm script called run_batch_predict.sh that calls batch_predict.py.

I'm sure you will need to adjust for your use case.  The key ideas to remember:

- the model config has to be the configuration saved in the checkpoint (not the model config from the pretrained model. If you used the caduceus fine tuning scripts then this would be located in the fine tuned model output directory ie ~/seed-1/config.json)
- the input has to be in csv file format (such as GUE) for this script to have the correct dataloading (ie sequence,label)
- I added some classes to monitor GPU usage so that I could choose the batch size appropriately.  You can comment this out if it is not needed for your project.

If you do end up using this code, I would appreciate it if you could cite our [paper](https://www.biorxiv.org/content/10.1101/2024.09.09.612081v1): 

A Comparison of Tokenization Impact in Attention Based and State Space Genomic Language Models
LeAnn M. Lindsey, Nicole L. Pershing, Anisa Habib, W. Zac Stephens, Anne J. Blaschke, Hari Sundar
bioRxiv 2024.09.09.612081; doi: https://doi.org/10.1101/2024.09.09.612081

You may also want to look at the [supplementary material](https://www.biorxiv.org/content/10.1101/2024.09.09.612081v1.supplementary-material?versioned=true), as it has some observations on sensitivity to batch size and learning rate that we saw during our experiments.

**Comment by JiaLonghao1997** [comment]
Thank you, I will try your code, and Happy New Year!
